airflow/exceptions.py

```diff
@@ -338,3 +338,17 @@ class TaskDeferralError(AirflowException):
 
 class PodReconciliationError(AirflowException):
     """Raised when an error is encountered while trying to merge pod configs."""
+
+
+class RemoveInAirflow3DeprecationWarning(DeprecationWarning):
```

Comment:
https://docs.python.org/3/library/warnings.html



DeprecationWarning | Base category for warnings about deprecated features when those warnings are intended for other Python developers (ignored by default, unless triggered by code in __main__).
-- | --
FutureWarning | Base category for warnings about deprecated features when those warnings are intended for end users of applications that are written in Python.
-- | --
PendingDeprecationWarning | Base category for warnings about features that will be deprecated in the future (ignored by default).

Should we make this baseclass be FutureWarning?

Reply:
We consider Airflow users as Python developers not end users don't we?

Reply:
`PendingDeprecationWarning`s and `DeprecationWarning`s are [filtered out](https://docs.python.org/3/library/warnings.html#default-warning-filter) by default. I think subclassing from `FutureWarning` would make them more visible to Airflow DAG authors.

Reply:
However, we may not want this to be immediately visible to DAG authors, because at this point we [should still assume that we will only have 2.x releases](https://apache-airflow.slack.com/archives/CCPRP7943/p1661211652517059).

Subclassing from DeprecationWarning (or PendingDeprecationWarning) might actually be the right call here until we have concrete plans for Airflow 3.0. [PEP-0387](https://peps.python.org/pep-0387/#making-incompatible-changes):

> If an API is being removed, simply warn whenever it is entered. DeprecationWarning is the usual warning category to use, but **PendingDeprecationWarning may be used in special cases where the old and new versions of the API will coexist for many releases [[2]](https://peps.python.org/pep-0387/#warnings)**.

(emphasis mine)

Reply:
We explicitly enable the Deprecation warnings coming from airflow following https://peps.python.org/pep-0565/#recommended-filter-settings-for-interactive-shells

See `airflow.configuration.py`:

```
if not sys.warnoptions:
    warnings.filterwarnings(action='default', category=DeprecationWarning, module='airflow')
    warnings.filterwarnings(action='default', category=PendingDeprecationWarning, module='airflow')
````

Reply:
I'm curious why #1285 "turned on" both [DeprecationWarning and PendingDeprecationWarning](https://github.com/apache/airflow/pull/1285/files#diff-2ab05df659c739ff335a407c4a6cddc5f177e9c402aca296991f53fa341f4dfeR20-R22).

Reply:
It was 6 years ago :) `5 Apr 2016` :). I am afraid no commiters who are active now were even around :D

Reply:
sorry to comment here, but i'm obsessed with DeprecationWarning recently because i couldn't find mine :joy: :
https://apache-airflow.slack.com/archives/CSS36QQS1/p1661977783108899?thread_ts=1661974104.498239&cid=CSS36QQS1